Chat-history: Add new buttons

just-carlod opened the pull request:
This pull request aim to improve chat-action-bar:
- [x] Implement mute button
- [x] Implement block button 
- [x] Implement join button

fixes: #307

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -109,5 +109,31 @@
+          <class name="blue-button"/>
+        </style>
+      </object>
+    </child>
```
It seems that these buttons can only be shown one at the time, right? I think that it makes sense to create a `GtkStack` that wraps everything inside the `GtkBox` and so creating multiple pages for each of the state. Also, this will make the buttons children of `GtkBox` + `.toolbar`, which will apply the correct style for them (6px margins + .flat).

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -165,6 +175,63 @@ mod imp {
+
+            // Mute/unmute channel
+            self.mute_button
+                .connect_clicked(clone!(@weak obj => move |_| {
```
Please use actions instead of signals to react to button press. You can look around for how it's done in other places.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -554,7 +621,71 @@ impl ChatActionBar {
+            matches!(args[1].get::<ChatType>().unwrap(), ChatType::Supergroup(data) if data.is_channel() && !matches!(data.status().0, ChatMemberStatus::Administrator(_) | ChatMemberStatus::Creator(_) | ChatMemberStatus::Left))
+        })
+            .upcast()
+            .bind(&*imp.mute_button, "visible", Some(chat));
```
All this bindings can probably be simplified by not using expressions, especially if you use a GtkStack. An example of how this can be reworked:

- Create a function to update the stack page based on the chat object properties.
- Connect notify signals from chat properties and make them call that update function.

just-carlod replied:
> Connect notify signals from chat properties and make them call that update function

Frankly speaking, I do not have any idea how to do this. Should I `notify_chat_connect` and listen for the change of the chat property? Or listen for the change of the Supergroup/BasicGroup/... etc?

melix99 replied:
> Frankly speaking, I do not have any idea how to do this

Yeah it's a bit complex, actually I'd like to use a new glib API to simplify this a bit. If you push the changes to fix the other reviews I can do it myself after this is merged, if you're ok with that.

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -225,6 +232,38 @@ impl ChatHistory {
 
+    fn leave_chat(&self) {
+        if let Some(chat) = self.chat() {
+            // TODO: Port this to adw::MessageDialog
```
Done :)

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -225,6 +232,37 @@ impl ChatHistory {
+        if let Some(chat) = self.chat() {
+            let dialog = adw::MessageDialog::new(
+                Some(&self.parent_window().unwrap()),
+                None,
```
```suggestion
                Some(gettext("Leave Chat?")),
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -1,7 +1,10 @@
+use adw::prelude::{MessageDialogExt, MessageDialogExtManual};
```
You can just change `use gtk::prelude::*;` to `use adw::prelude::*;`.

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -13,6 +16,7 @@ const MIN_N_ITEMS: u32 = 20;
 mod imp {
     use super::*;
     use adw::subclass::prelude::BinImpl;
+
```
```suggestion
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -225,6 +232,37 @@ impl ChatHistory {
         }
     }
 
+    fn leave_chat(&self) {
```
```suggestion
    fn show_leave_chat_dialog(&self) {
```

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+            <property name="child">
+              <object class="GtkButton" id="join_button">
+                <property name="action-name">chat-action-bar.join-chat</property>
+                <property name="hexpand">True</property>
```
```suggestion
```

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+            <property name="child">
+              <object class="GtkButton" id="mute_button">
+              <property name="action-name">chat-action-bar.mute-chat</property>
+              <property name="hexpand">True</property>
```
```suggestion
```

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+            <property name="child">
+              <object class="GtkButton" id="unblock_button">
+                <property name="action-name">chat-action-bar.unblock-chat</property>
+                <property name="hexpand">True</property>
```
```suggestion
```

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+            <property name="child">
+              <object class="GtkLabel" id="restriction_label">
+                <property name="height-request">48</property>
+                <property name="hexpand">True</property>
```
```suggestion
```

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+            <property name="name">restricted</property>
+            <property name="child">
+              <object class="GtkLabel" id="restriction_label">
+                <property name="height-request">48</property>
```
```suggestion
```

I'm not 100%, but I think it's not needed anymore since we have a `GtkStack` that already allocated the right height.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,53 @@ mod imp {
                 },
             );
+
+            klass.install_action("chat-action-bar.join-chat", None, move |widget, _, _| {
```
You can use the new `install_action_async`, this way you don't need to call `spawn` anymore.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,53 @@ mod imp {
+                })
+            });
+
+            klass.install_action("chat-action-bar.mute-chat", None, move |widget, _, _| {
```
Same here.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,53 @@ mod imp {
+                }));
+            });
+
+            klass.install_action("chat-action-bar.unblock-chat", None, move |widget, _, _| {
```
Same here.

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -225,6 +232,37 @@ impl ChatHistory {
+            dialog.set_default_response(Some("no"));
+            dialog.set_close_response("no");
+            dialog.set_response_appearance("yes", adw::ResponseAppearance::Destructive);
+            dialog.run_async(
```
I'd make this use `run_future` and I'd make the entire `show_leave_chat_dialog` function async. This way the `spawn` below can be removed.

It should look something like:

```rust
let response = dialog.run_future(None).await;
dialog.close();

let result = ...
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,53 @@ mod imp {
+                    if widget.is_chat_muted() {
+                        btn.set_label("Unmute");
+                    } else {
+                        btn.set_label("Mute");
```
These strings should be wrapped in `gettext`.

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,108 @@
+                <property name="hexpand">True</property>
+                <property name="label">Join</property>
+                <style>
+                </style>
```
Needs `.flat` and `.squared`.

just-carlod commented on src/session/content/chat_history.rs:
```diff
@@ -279,6 +317,14 @@ impl ChatHistory {
+                matches!(
+                    chat.type_(),
+                    ChatType::Supergroup(_) | ChatType::BasicGroup(_)
+                ),
```
Leave action shouldn't be enabled if the user is not a member of the chat

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -65,47 +65,106 @@
+            <property name="name">mute</property>
+            <property name="child">
+              <object class="GtkButton" id="mute_button">
+              <property name="action-name">chat-action-bar.mute-chat</property>
```
```suggestion
              <property name="action-name">chat-action-bar.toggle-mute</property>
```

It's more descriptive of what it actually does.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,54 @@ mod imp {
             );
+
+            klass.install_action_async("chat-action-bar.join-chat", None, move |widget, _, _| {
+                async move {
```
```suggestion
            klass.install_action_async("chat-action-bar.join-chat", None, |widget, _, _|
                async move {
```

No need for the extra block.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,54 @@ mod imp {
+            klass.install_action_async(
+                "chat-action-bar.mute-chat",
+                None,
+                move |widget, _, _| async move {
```
```suggestion
                |widget, _, _| async move {
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -78,6 +85,54 @@ mod imp {
+
+            klass.install_action_async("chat-action-bar.unblock-chat", None, move |widget, _, _| {
+                let chat = widget.chat().unwrap();
+                async move {
```
```suggestion
            klass.install_action_async("chat-action-bar.unblock-chat", None, |widget, _, _| async move {
                let chat = widget.chat().unwrap();
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -225,6 +236,31 @@ impl ChatHistory {
         }
     }
 
+    async fn should_leave_chat_dialog(&self) {
```
```suggestion
    async fn show_leave_chat_dialog(&self) {
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -279,6 +315,19 @@ impl ChatHistory {
+                "chat-history.leave-chat",
+                match chat.type_() {
+                    ChatType::BasicGroup(data) => {
+                        !matches!(data.status().0, ChatMemberStatus::Left)
```
```suggestion
                        data.status().0 != ChatMemberStatus::Left
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -279,6 +315,19 @@ impl ChatHistory {
+                        !matches!(data.status().0, ChatMemberStatus::Left)
+                    }
+                    ChatType::Supergroup(data) => {
+                        !matches!(data.status().0, ChatMemberStatus::Left)
```
```suggestion
                        data.status().0 != ChatMemberStatus::Left
```

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -1,7 +1,11 @@
+use gettextrs::gettext;
 use glib::clone;
-use gtk::prelude::*;
+
```
```suggestion
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -575,6 +577,84 @@ impl ChatActionBar {
+                    let is_blocked = chat.is_blocked();
+                    if is_deleted {
+                        // TODO: Add delete chat button
+                        self.imp().action_bar_stack.set_visible_child_name("entry");
```
To save quite a bit of code, can you put `let imp = self.imp();` at the start of the function and then use that instead of using `self.imp()` every time? That's a pattern we use everywhere.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -45,6 +44,14 @@ mod imp {
+        #[template_child]
+        pub(super) mute_button: TemplateChild<gtk::Button>,
+        #[template_child]
+        pub(super) join_button: TemplateChild<gtk::Button>,
```
This seems unused.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -45,6 +44,14 @@ mod imp {
         #[template_child]
         pub(super) restriction_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub(super) unblock_button: TemplateChild<gtk::Button>,
```
This too.

melix99 reviewed (changes requested):
Last batch of changes before the merge.

krlade committed:
chat-history: Add new buttons

Added join, mute and block button to the ChatActionBar and the leave
button to the chat headerbar menu.

Co-authored-by: Marco Melorio <marco.melorio@protonmail.com>
